File: src/retraite_notionnelle/moteur/conversion.py
"""Coefficient de conversion du capital notionnel en rente viagère.

La pension annuelle vaut ``capital_notionnel / diviseur``. Le diviseur est
l'espérance de vie résiduelle actualisée :

.. math::

   G(a, L) = \\sum_{t \\ge 0} \\; {}_t p_a \\; (1 + \\nu)^{-t}

où :math:`{}_t p_a` est la probabilité, pour un liquidant d'âge :math:`a` en
année :math:`L`, d'être encore en vie :math:`t` années plus tard, lue sur une
table de **génération**, et :math:`\\nu` le taux de préfinancement.

**Pourquoi :math:`\\nu = 0` par défaut.** Dans un système notionnel, la rente
est actualisée au taux auquel elle sera ensuite revalorisée. Ici les deux sont
le même taux — le triple lock inversé — et se compensent exactement. Le diviseur
se réduit alors à l'espérance de vie résiduelle, ce qui rend le résultat
directement lisible : « votre capital notionnel divisé par le nombre d'années
que vous êtes statistiquement appelé à vivre ». Donner à :math:`\\nu` une valeur
positive revient à verser davantage au début et moins ensuite, à espérance de
coût inchangée.

**Ce que le diviseur sanctionne tout seul.** Partir cinq ans plus tôt augmente
le diviseur d'environ 4 à 5 années d'espérance de vie, soit une pension annuelle
inférieure de 15 à 20 % — avant même de compter les cinq années de cotisations
manquantes. C'est ce mécanisme qui traduit la règle « parti trop tôt, pension
réduite » sans avoir besoin d'une décote administrative.
"""

from __future__ import annotations

from dataclasses import dataclass

from ..config import Parametres, TableConversion
from ..donnees.chargement import Fiabilite
from ..donnees.mortalite import DonneesMortalite


@dataclass(frozen=True)
class CoefficientConversion:
    """Diviseur annuitaire et éléments qui l'expliquent."""

    diviseur: float
    age_liquidation: float
    annee_liquidation: int
    esperance_residuelle: float
    table: str
    taux_anticipe: float
    fiabilite: Fiabilite

    @property
    def taux_de_rente(self) -> float:
        """Fraction du capital notionnel servie chaque année."""
        return 1.0 / self.diviseur if self.diviseur else 0.0


class Convertisseur:
    """Produit les coefficients de conversion."""

    def __init__(self, mortalite: DonneesMortalite, parametres: Parametres) -> None:
        self.mortalite = mortalite
        self.parametres = parametres

    def _sexe_table(self, sexe: str | None) -> str | None:
        if self.parametres.table_conversion is TableConversion.UNISEXE:
            return None
        if sexe is None:
            raise ValueError(
                "table de conversion par sexe demandée mais sexe non renseigné"
            )
        return sexe

    def coefficient(self, age_liquidation: float, annee_liquidation: int,
                    sexe: str | None = None) -> CoefficientConversion:
        sexe_table = self._sexe_table(sexe)
        generation = self.parametres.table_generation
        courbe = self.mortalite.courbe(
            age_liquidation, annee_liquidation, sexe_table, generation
        )

        nu = self.parametres.taux_anticipe_conversion
        diviseur = 0.0
        for t in range(len(courbe) - 1):
            # Rente supposée servie en continu sur l'année : on prend la survie
            # moyenne de début et de fin de période.
            survie_moyenne = 0.5 * (courbe[t] + courbe[t + 1])
            diviseur += survie_moyenne / ((1.0 + nu) ** (t + 0.5))

        if diviseur <= 0:
            raise ValueError(
                f"diviseur nul à {age_liquidation} ans en {annee_liquidation} : "
                "âge de liquidation hors des bornes de la table"
            )

        esperance = sum(
            0.5 * (courbe[t] + courbe[t + 1]) for t in range(len(courbe) - 1)
        )
        return CoefficientConversion(
            diviseur=diviseur,
            age_liquidation=age_liquidation,
            annee_liquidation=annee_liquidation,
            esperance_residuelle=esperance,
            table=("unisexe" if sexe_table is None else sexe_table)
            + ("_generation" if generation else "_moment"),
            taux_anticipe=nu,
            fiabilite=self.mortalite.fiabilite(annee_liquidation),
        )

    def effet_anticipation(self, age_anticipe: float, age_reference: float,
                           annee_liquidation: int, sexe: str | None = None) -> float:
        """Rapport des pensions à capital notionnel donné, anticipé / à l'heure.

        Isole la seule sanction due à l'allongement de la durée de service.
        L'effet total d'un départ anticipé est plus fort, puisque s'y ajoutent
        les cotisations non versées.
        """
        anticipe = self.coefficient(age_anticipe, annee_liquidation, sexe)
        reference = self.coefficient(age_reference, annee_liquidation, sexe)
        return reference.diviseur / anticipe.diviseur

File: moteur/js/conversion.js
/**
 * Coefficient de conversion du capital notionnel en rente viagère.
 *
 * Portage de ``src/retraite_notionnelle/moteur/conversion.py``. La pension
 * annuelle vaut ``capital_notionnel / diviseur``, où le diviseur est
 * l'espérance de vie résiduelle actualisée :
 *
 *     G(a, L) = Σ_t  t·p_a  (1 + ν)^(−t)
 *
 * lue sur une table de génération. Le taux de préfinancement ν vaut 0 par
 * défaut : la rente est actualisée au taux auquel elle sera revalorisée, les
 * deux se compensent, et le diviseur se réduit à l'espérance de vie résiduelle.
 *
 * Ce que le diviseur sanctionne tout seul : partir cinq ans plus tôt l'augmente
 * de quatre à cinq années, soit une pension inférieure de 15 à 20 % — avant
 * même de compter les cinq années de cotisations manquantes.
 */

import { TableConversion } from "./config.js";

export class Convertisseur {
  constructor(mortalite, parametres) {
    this.mortalite = mortalite;
    this.parametres = parametres;
  }

  _sexeTable(sexe) {
    if (this.parametres.table_conversion === TableConversion.UNISEXE) {
      return null;
    }
    if (sexe === null || sexe === undefined) {
      throw new Error("table de conversion par sexe demandée mais sexe non renseigné");
    }
    return sexe;
  }

  /** Diviseur annuitaire et éléments qui l'expliquent. */
  coefficient(ageLiquidation, anneeLiquidation, sexe = null) {
    const sexeTable = this._sexeTable(sexe);
    const generation = this.parametres.table_generation;
    const courbe = this.mortalite.courbe(
      ageLiquidation, anneeLiquidation, sexeTable, generation,
    );

    const nu = this.parametres.taux_anticipe_conversion;
    let diviseur = 0.0;
    for (let t = 0; t < courbe.length - 1; t += 1) {
      // Rente supposée servie en continu sur l'année : on prend la survie
      // moyenne de début et de fin de période.
      const survieMoyenne = 0.5 * (courbe[t] + courbe[t + 1]);
      diviseur += survieMoyenne / ((1.0 + nu) ** (t + 0.5));
    }

    if (diviseur <= 0) {
      throw new Error(
        `diviseur nul à ${ageLiquidation} ans en ${anneeLiquidation} : `
        + "âge de liquidation hors des bornes de la table",
      );
    }

    let esperance = 0.0;
    for (let t = 0; t < courbe.length - 1; t += 1) {
      esperance += 0.5 * (courbe[t] + courbe[t + 1]);
    }

    return {
      diviseur,
      age_liquidation: ageLiquidation,
      annee_liquidation: anneeLiquidation,
      esperance_residuelle: esperance,
      table: (sexeTable === null ? "unisexe" : sexeTable)
        + (generation ? "_generation" : "_moment"),
      taux_anticipe: nu,
      fiabilite: this.mortalite.fiabilite(anneeLiquidation),
      /** Fraction du capital notionnel servie chaque année. */
      get taux_de_rente() {
        return this.diviseur ? 1.0 / this.diviseur : 0.0;
      },
    };
  }

  /**
   * Rapport des pensions à capital notionnel donné, anticipé / à l'heure.
   * Isole la seule sanction due à l'allongement de la durée de service.
   */
  effetAnticipation(ageAnticipe, ageReference, anneeLiquidation, sexe = null) {
    const anticipe = this.coefficient(ageAnticipe, anneeLiquidation, sexe);
    const reference = this.coefficient(ageReference, anneeLiquidation, sexe);
    return reference.diviseur / anticipe.diviseur;
  }
}
